여러 작업을 함에 있어서 이전의 작업이 끝나면 시작하는 것> 동기, 모두 동시에 일을 시작하는것 비동기;



promise > callback 기능을 가진 함수

callback이란 ? 함수가 끝나고 실행시킬 함수;

```
function asd(q){
	return new count = new Promise(resolve, reject){
		if(q < 5){
			resolve(a); >>promise가 성공하면 a를 반환	
		}else{
			b= new Error();
			b.name = "5보다같거나큼"
			reject(b); >> 실패하면 b를 반환
		}
	}
}

then, catch 문법 > 비동기 처리의 결과에 따라서 결과값을 반환
promise 의 성공은 resolve, 실패는 reject; then은 성공, catch는 실패


asd(4).then( function(parms){
	console.log(parms);
}).catch( (parms){
	console.log(parms)
});
```

async await

```
function ser(awe){
	return new Promise( function(res){
		setTimeOut( res, awe);
	}) 
}

async function gogo(){
	console.log("dsds")
	await ser(2000);
	console.log("asdds");
	return "dd"
}
```



질문 1 . callback의 경우 왜 파라미터 딴에서 저장을 하는가,

```
function bedonggi(n, callback){
	setTimeOut(function(){
		nax= n+1
		console.log(nax);
		if(callback){
			callback(nax)
		}
	},2000)
}

> 2초뒤에 파라미터 n에 1을 더하고 콘솔에 출력, 파라미터 콜백자리에 콜백이 있으면 콜백함수에 nax를 파라미터로 넣고 실행

bedonggi(0,function(n){
// 2초뒤에 콘솔에 1 출력 / 0 +1 이니까 그리고 nax에 1이 들어감.
	bedonggi(n,function(){
	//콜백의 파라미터에 nax가 들어가니 1+1 콘솔에 2가 출력
		console.log("끝")
	})
})
```

```
const myP = new Promise(function(res,rej){
	for( var i =0; i < 5; i++){
		setTimeout(function(i){
			console.log("i")
		},2000);
	res("success")
	}
});
```

결국 promise의 res, rej는 콜백 함수인거다.